File: Binary Tree/二叉树的属性/104.二叉树的最大深度.cpp
#include <queue>

using namespace std;
/*
 * @lc app=leetcode.cn id=104 lang=cpp
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// BFS解决
// class Solution
// {
// public:
//     int maxDepth(TreeNode *root)
//     {
//         int level = 0;
//         queue<TreeNode *> qu;

//         if (root != nullptr)
//         {
//             qu.push(root);
//         }

//         while (!qu.empty())
//         {
//             int size = qu.size();
//             level++;
//             for (int i = 0; i < size; i++)
//             {
//                 TreeNode *temp = qu.front();
//                 qu.pop();

//                 if (temp->left != nullptr)
//                 {
//                     qu.push(temp->left);
//                 }
//                 if (temp->right != nullptr)
//                 {
//                     qu.push(temp->right);
//                 }
//             }
//         }

//         return level;
//     }
// };
// 递归
class Solution
{
public:
    int getLevel(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int left = getLevel(root->left);
        int right = getLevel(root->right);

        return 1 + max(left, right);
    }
    int maxDepth(TreeNode *root)
    {
        return getLevel(root);
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/104.二叉树的最大深度.ts
/*
 * @lc app=leetcode.cn id=104 lang=typescript
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxDepth(root: TreeNode | null): number {
  let queue = [root]
  let level = 0

  if (!root) {
    return level
  }

  while (queue.length > 0) {
    // 当前层级的左右节点数量
    let size = queue.length

    // 移除当前层级节点，移除出的同时将子节点放入队列
    for (let i = 0; i < size; i++) {
      let node = queue.shift()

      if (node) {
        node.left && queue.push(node.left)
        node.right && queue.push(node.right)
      }
    }
    // 遍历完一层，层级+1
    level++
  }

  return level
};
// @lc code=end

